user guide: infrastructure

oliver-sanders opened the pull request:
The first part of the rose user guide overhaul, this PR lays the software foundation on top of which the tutorials will be written.

__Highlights__

* Remove old documentation
* New documentation written in sphinx/ and built by sphinx to doc/
* Custom sphinx extensions written
* Temporary demo section added for assessment

__Rose Commands__

Transitioned away from using a Makefile to build the documentation to using two rose commands:
* `rose make-docs [BUILDER ..]`
* `rose check-software-docs [BUILDER ..]`

The `rose make-docs` command will check software dependencies prior to building. If they are insufficient it offers the user the option of installing dependencies in a virtual environment. The virtual environment will be cleaned after use unless either the `--dev` option is supplied or the `SPHINX_DEV_MODE` environment variable is provided (for use with the `rose test-battery` command).

For development run `rose make-docs --dev strict linkcheck doctest`, this will check external links, run doctests and will fail if any warnings occur in the build, this is what is run in the `rose test-battery`.

__Sphinx Plugins__

A few plugins have been developed to make documentation faster to write and more standardised. These are implemented as ReStructured text directives.

* Rose lexer - _pygments lexer for rose config files_.
* Cylc lexer - _pygments lexer for suite.rc files_.
* Practical directive - _a document section for writing practical sections, (the purpose of this is for standardisation)._
* minicylc directive - _turns cylc graph strings into animated SVG graphs_.
* auto-cli-doc directive - _autodocuments the rose/rosie command line API_.


__minicylc__

The minicylc directive generates a SVG graph from cylc graph strings. When built to HTML these graphs are animated in-browser using javascript. At present all tasks run for the same period and all succeed. Conditional statements are enabled but cycling, qualifier dependencies (e.g. `foo:fail`) and parametrisation are not supported. _If it is considered to be a worthwhile investment these could be added._

__Auto CLI API Documentation__

The rose/rosie commands are now auto-documented. This is achieved by parsing the command line help and writing out ReStructured Text.

* The `NAME` section will be ignored.
* The `SYNOPSIS` section will be displayed as code with bash syntax highlighting.
* Any `OPTIONS`, `ARGUMENTS`, `ENVIRONMENT_VARIABLES`, `JINJA2 VARIABLES`, `CONFIGURATIONS` sections will be treated as option lists. The options will be rendered as literals (code) and the descriptions as RST.
* Any other sections will be rendered as RST.

This means that the rose help is now written in (RST) markup language and must be standardised. Note bullet points / lists must be preceded and followed by blank lines and three space indentation must be used. RST literals (code) is denoted using double backquotes, to keep the documentation clean single backquoted text is written out to sphinx as double back quotes. Any command name written in single back-qoutes will result in a hyperlink to the documentation section for that command E.G:
```
see also `rose task-env`
```
Definition lists are particularly useful here and are written like so:
```
Item

   Description

   All indented paragraphs are part of the description:

   * Note the three space indentation
   * And newline preceding / following each section.

Next Item

   Next description.
```
Code examples can be displayed using literal blocks (as opposed to using the `code-block` directive as we would in the tutorial material:
```
Lorel ipsum::

   mono space dolor imet epit

```

__Theme__

The sphinx theme has been switched to the popular readthedocs theme. Syntax highlighting was a small challenge due to the large number of un-conventional roles present in the exotic rose/cylc configuration files, consequently the default syntax highlighting has been overwritten to use the pygments built-in theme `autumn`. Note that due to a bug in the sphinx readthedocs theme (https://github.com/rtfd/sphinx_rtd_theme/issues/166) the syntax hightlighter cannot be changed hence the `Fix read the docs syntax highlighting` commit.

It is possible to generate PDF via LaTeX using `rose make-docs latexpdf`, the produced PDF is usable but could do with prettifying. URLs will be displayed in brackets next to the anchor text, minicylc blocks will be displayed as blank graphs (no animation obviously).

__TODO__

- This PR does not cover versioning of documentation, this is something we should look into, there is a sphinx plugin which may make this easier - https://github.com/sphinxcontrib-versioning.
- PDF output (via LaTeX) has a lot to be desired.
- Write the documentation :)

__Try Before You Buy__
```bash
$ rose make-docs html latexpdf
$ firefox doc/html/index.html doc/latex/rose-documentation.pdf
```

oliver-sanders committed:
Added cylc lexer and builtin plugins.
* Added experimental Minicylc directive.
* Chaged to RTD theme.
* Updated latex options.
* Changed syntax highlighting theme.

oliver-sanders committed:
Added 'practical' directive.

oliver-sanders committed:
MiniCylc directive.

oliver-sanders committed:
Fix read the docs syntax highlighting.

oliver-sanders committed:
Added rose config file language lexer.

oliver-sanders committed:
Deleted old docs.

oliver-sanders committed:
Moved sphinx directory (changed apidocs to docs).

oliver-sanders committed:
Added rose command to make documentation.

oliver-sanders committed:
Added rose-version file.

oliver-sanders committed:
Added demo section.

oliver-sanders committed:
Show urls in latexpdf.

oliver-sanders committed:
Added CLI auto-documentation.

oliver-sanders committed:
Updated test-battery.

oliver-sanders committed:
pep8.

oliver-sanders committed:
Add cylc graph lexer for code snippets.

oliver-sanders committed:
Remove config TODOs.

oliver-sanders committed:
Resolve ESLint issues.

oliver-sanders committed:
Add pyopenssl to travis environment.

oliver-sanders committed:
Add eslint javascript syntax check.

oliver-sanders committed:
Fix rose-suite-run test.

oliver-sanders commented:
Following on from https://github.com/metomi/rose/pull/2046#issuecomment-288132630 Travis-ci now offers a built-in [deployment system](https://docs.travis-ci.com/user/deployment/pages/) for github pages...

matthewrmshin commented on bin/rose:
```diff
@@ -151,6 +151,33 @@ version|--version|-V)
+            echo "$ROSE_NS $NAME"
+            echo '=================================================='
+            echo
+            PAGER=cat
```
Does this need `export`?

oliver-sanders replied:
Apparently not...

matthewrmshin commented on bin/rose:
```diff
@@ -151,6 +151,33 @@ version|--version|-V)
+            echo '=================================================='
+            echo
+            PAGER=cat
+            help_util $(echo $U | cut -c ${ns_len}-)
```
This is more efficient: `cut -c "${ns_len}-" <<<"${U}"`.

matthewrmshin commented on bin/rose-check-software-docs:
```diff
@@ -0,0 +1,200 @@
+        if max_version and not version < max_version:
+            ret = False
+    else:
+        ret = False
```
`ret = version < min_version or max_version and version >= max_version`?

matthewrmshin commented on bin/rose-check-software-docs:
```diff
@@ -0,0 +1,200 @@
+    else:
+        ret = False
+    sys.stdout.write('%s (%s)\n' % ('PASS' if ret else 'FAIL -',
+                                      string_version(version)))
```
If `PASS` and `FAIL` are not special format, then it is worth using TAP, e.g. `ok` and `not ok` instead.

matthewrmshin commented on bin/rose-check-software-docs:
```diff
@@ -0,0 +1,200 @@
+    if all(ret):
+        sys.exit(0)
+    else:
+        sys.exit(1)
```
`sys.exit(not all(ret))`

matthewrmshin commented on bin/rose-app-upgrade:
```diff
@@ -12,8 +12,9 @@
-#         "*" indicates the default version to change to.
+#
+#     * "=" indicates the current version.
+#     * "*" indicates the default version to change to.
```
Back tick the equal and star as well?

oliver-sanders replied:
I had left that purposely, I'll change it though.

matthewrmshin commented on bin/rose-make-docs:
```diff
@@ -0,0 +1,175 @@
+#-----------------------------------------------------------------------------
+
+# Move into rose directory
+cd $(dirname $0)/../
```
We should always quote path strings in shell script. `cd "$(dirname "$0")/../"`

Note: I know I set a bad example elsewhere, but I am trying to amend that these days...

oliver-sanders replied:
Yes, that line was copied! I still should have spotted it though.

matthewrmshin commented on bin/rose-make-docs:
```diff
@@ -0,0 +1,175 @@
+VENV=false  # Set to `true` when the virtualenv is being used.
+SPHINX_PATH=sphinx  # Path to the sphinx directory.
+
+function venv-activate () {
```
The function keyword is redundant in most usage.

Note: We should aim to remove them from `bin/rose` as well, but new script function should be written without the function keyword.

matthewrmshin commented on bin/rose-make-docs:
```diff
@@ -0,0 +1,175 @@
+            ;;
+    esac
+done
+if [[ ${BUILDS} == '' ]]; then
```
`if [[ -z "${BUILDS}" ]]; then`

matthewrmshin commented on bin/rose-make-docs:
```diff
@@ -0,0 +1,175 @@
+fi
+
+# Use virtualenv if present and requested or if we are in development mode.
+if [[ ${DEV_MODE} || ${YES} ]]; then
```
`if "${DEV_MODE:-false}" || "{YES:-false}"; then`

matthewrmshin commented on bin/rose-make-docs:
```diff
@@ -0,0 +1,175 @@
+#         Checks external links.
+#
+#-----------------------------------------------------------------------------
+
```
`set -eu` and `set -o pipefail`

oliver-sanders replied:
Can do `set -e`, cant do `set -u` (would break the code called in `venv-install`).

matthewrmshin replied:
The only line I think would break is:

`virtualenv --python=python2.7 "${VENV_PATH}"`

You can do:

`virtualenv --python=python2.7 "${VENV_PATH:-}"`

matthewrmshin commented on bin/rose-make-docs:
```diff
@@ -0,0 +1,175 @@
+        echo 'with the "--dev" option.'
+        echo
+        while true; do
+            read -p 'Proceed using a virtualenv (y/n)? ' in
```
You can do `while read ...; do`. Also it is best not to use `in` as the variable.

matthewrmshin commented on sphinx/ext/rose-lang.py:
```diff
@@ -0,0 +1,150 @@
+from pygments.lexer import RegexLexer, bygroups, include, words
```
Add header.

matthewrmshin commented on sphinx/ext/rose-lang.py:
```diff
@@ -0,0 +1,150 @@
+            (r'\[.*\]', Name.Tag, 'section'),
+        ],
+
+        # Rose comments - single and inline w/ or w/o/ leading whitespace.
```
Rose configuration file does not have inline comment.

matthewrmshin commented on sphinx/ext/rose-lang.py:
```diff
@@ -0,0 +1,150 @@
+    ROSE_SETTING_PATTERN = r'^([\s\t]+)?([\w\{\}-]+)(\s+)?(=)(\s+)?'
+    ROSE_SETTING_VALUE_PATTERN = ('{0}'  # '', '!', '!!'.
+                                  r'[\w{{}}-]+(\s+)?=(\s+)?'
+                                  r'[^\n]+((\n[\s\t]+=[^\n]+)+)?')
```
Maybe worth adding some extra comments to the regular expressions.

matthewrmshin commented on sphinx/ext/practical.py:
```diff
@@ -0,0 +1,25 @@
+from docutils import nodes
```
Add header.

matthewrmshin commented on sphinx/ext/minicylc.py:
```diff
@@ -0,0 +1,231 @@
+"""Provides the minicylc directive and its associated node as sphinx
```
Add header.

matthewrmshin commented on sphinx/ext/cylc-lang.py:
```diff
@@ -0,0 +1,232 @@
+from pygments.lexer import RegexLexer, bygroups, include, words
```
Add header.

matthewrmshin commented on sphinx/ext/cylc-lang.py:
```diff
@@ -0,0 +1,232 @@
+        'parameterisation': [
+            (r'(\<)(\s?\w+\s?(?:[+-=]\s?\d+)?\s?'
+             r'(?:(?:\s?,\s?\w+\s?(?:[+-=]\s?\d+)?\s?)+)?'
+             r')(\>)',
```
It is worth having a bit more comment for the regular expression here.

matthewrmshin commented on sphinx/ext/auto-cli-doc.py:
```diff
@@ -0,0 +1,377 @@
+"""An extension for auto-documenting command line interfaces."""
```
Add header.

matthewrmshin commented on sphinx/conf.py:
```diff
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
```
Add header.

matthewrmshin reviewed (commented):
Some initial comment. Mainly on coding style. I'll have a good look at the directory structure, etc next.

oliver-sanders commented on bin/rose-make-docs:
```diff
@@ -0,0 +1,175 @@
+#         Checks external links.
+#
+#-----------------------------------------------------------------------------
+
```
Script now brexit ready!

oliver-sanders committed:
Response to feedback.

dvalters commented on bin/rose-make-docs:
```diff
@@ -0,0 +1,176 @@
+        echo 'present (run `rose software-check-docs` for details).'
+        echo
+        echo 'The documentation can still be built by installing dedencies'
+        echo 'in a python virtual enironment (virtualenv). This environment'
```
typo 'dependencies'

dvalters commented on bin/rose-make-docs:
```diff
@@ -0,0 +1,176 @@
+}
+
+# Parse command line args.
+YES=false
```
Could it be called something like `VENV_MODE` for clarity?

dvalters commented on bin/rose-make-docs:
```diff
@@ -0,0 +1,176 @@
+cd "$(dirname $0)/../"
+
+VENV_PATH='venv'  # Path for virtualenv.
+VENV=false  # Set to `true` when the virtualenv is being used.
```
Does this variable do the same thing as `YES` flag? (if not it could be something like `VENV_IN_USE`)

dvalters reviewed (commented):
Looks really good so far - I like the general style and all the sphinx functionality (admonitions etc.). Some initial comments so far.

dvalters commented on sphinx/ext/cylc-lang.py:
```diff
@@ -0,0 +1,261 @@
+strings."""
+
+from pygments.lexer import RegexLexer, bygroups, include, words
+from pygments.token import *
```
Better to import specific names rather than whole namespace? (or `import pygments.token as pt` . (I thought this was in PEP 8 but apparently its not, maybe it's just a convention.)

oliver-sanders replied:
Agreed, It's certainly in pylint, don't know about pep8. I'd left this has a star import as there are quite a lot of namespaces to import!

dvalters commented on sphinx/ext/rose-lang.py:
```diff
@@ -0,0 +1,173 @@
+"""An extension providing a pygments lexer for rose configuration files."""
+
+from pygments.lexer import RegexLexer, bygroups, include, words
+from pygments.token import *
```
same as previous star import

dvalters commented on bin/rose-make-docs:
```diff
@@ -0,0 +1,176 @@
+            shift
+            ;;
+        *)
+            BUILDS="$BUILDS $1"
```
Do you need to curly-brace `{BUILDS}` here (more for consistency really..)

dvalters commented on bin/rose-make-docs:
```diff
@@ -0,0 +1,176 @@
+
+# Check core (sphinx) builder.
+if ! rose-check-software-docs core >/dev/null; then
+    if ${VENV_MODE}; then
```
Same with as above, I'm not entirely sure of the convention here, but other bash scripts seem to quote plus curly brace (even if not strictly necessary??)

oliver-sanders replied:
I've not used quotes here as `VENV_MODE` is a constant which is set to either `true` or `false`, quotes defend against spaces which be erroneous if present.

dvalters commented on bin/rose-make-docs:
```diff
@@ -0,0 +1,176 @@
+    fi
+fi
+
+# Run sphinx make.
```
Variable quoting consistency below here:

oliver-sanders replied:
This is purposeful, the `BUILDS` variable contains space separated variables which should be interpreted as arguments rather than passing the whole variable `BUILD` as a single argument.

dvalters commented on bin/rose-make-docs:
```diff
@@ -0,0 +1,176 @@
+for build in ${BUILDS}; do
+    if ! rose-check-software-docs ${build} >/dev/null; then
+        echo "Software dependencies for a ${build} build are missing."
+        echo "run `rose software-check-docs ${build}` for details."
```
It should be `rose check-software-docs` I think

oliver-sanders replied:
Yes it should be! I renamed the command.

dvalters commented on bin/rose-make-docs:
```diff
@@ -0,0 +1,176 @@
+    else
+        echo 'Software required by the rose documentation builder is not'
+        echo 'present (run `rose software-check-docs` for details).'
+        echo
```
`rose check-software-docs`

dvalters commented on bin/rose-make-docs:
```diff
@@ -0,0 +1,176 @@
+#         For building a single page HTML document.
+#     `latexpdf`
+#         For building PDF documentation.
+#
```
The `clean` command might be useful to document here (just so users know they can clean the docs.)

oliver-sanders replied:
Good idea.

dvalters reviewed (commented):
Build options working well when tested.

Some more comments (mostly small):

Maybe the sphinx directory could go into docs? (then all the documentation stuff is in one place.)

oliver-sanders commented:
Addressed feedback, good spot on the `rose software-check-docs`!

> Maybe the sphinx directory could go into docs?

This is how it was initially, I moved it out largely for deployment reasons. This way when built the `doc` directory is all we need to push up to the gh-pages branch.

dvalters reviewed (approved):
Great stuff!
